File: src/test/java/de/pflugradts/passbird/domain/service/password/PutPasswordServiceTest.java
package de.pflugradts.passbird.domain.service.password;

import de.pflugradts.passbird.application.eventhandling.PassbirdEventRegistry;
import de.pflugradts.passbird.application.security.CryptoProviderFaker;
import de.pflugradts.passbird.domain.model.Tuple;
import de.pflugradts.passbird.domain.model.password.InvalidKeyException;
import de.pflugradts.passbird.domain.model.password.PasswordEntry;
import de.pflugradts.passbird.domain.model.password.PasswordEntryFaker;
import de.pflugradts.passbird.domain.model.password.PasswordEntryRepositoryFaker;
import de.pflugradts.passbird.domain.model.transfer.Bytes;
import de.pflugradts.passbird.domain.service.NamespaceServiceFake;
import de.pflugradts.passbird.domain.service.password.encryption.CryptoProvider;
import de.pflugradts.passbird.domain.service.password.storage.PasswordEntryRepository;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Spy;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.stream.Stream;

import static de.pflugradts.passbird.domain.model.namespace.NamespaceSlot.DEFAULT;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.catchThrowable;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.BDDMockito.then;

@ExtendWith(MockitoExtension.class)
class PutPasswordServiceTest {

    @Mock
    private CryptoProvider cryptoProvider;
    @Spy
    private final NamespaceServiceFake namespaceServiceFake = new NamespaceServiceFake();
    @Mock
    private PasswordEntryRepository passwordEntryRepository;
    @Mock
    private PassbirdEventRegistry passbirdEventRegistry;
    @InjectMocks
    private PutPasswordService passwordService;

    @Test
    void shouldSucceedChallengingAlphabeticAlias() {
        // given
        final var givenAlias = Bytes.bytesOf("abcDEF");

        // when
        final var actual = catchThrowable(() -> passwordService.challengeAlias(givenAlias));

        // then
        assertThat(actual).isNull();
    }

    @Test
    void shouldSucceedChallengingAliasWithDigitAtOtherThanFirstPosition() {
        // given
        final var givenAlias = Bytes.bytesOf("abc123");

        // when
        final var actual = catchThrowable(() -> passwordService.challengeAlias(givenAlias));

        // then
        assertThat(actual).isNull();
    }

    @Test
    void shouldFailChallengingAliasWithDigitAtFirstPosition() {
        // given
        final var givenAlias = Bytes.bytesOf("123abc");

        // when
        final var actual = catchThrowable(() -> passwordService.challengeAlias(givenAlias));

        // then
        assertThat(actual).isNotNull().isInstanceOf(InvalidKeyException.class);
    }


    @Test
    void shouldFailChallengingAliasWithSpecialCharacters() {
        // given
        final var givenAlias = Bytes.bytesOf("abc!");

        // when
        final var actual = catchThrowable(() -> passwordService.challengeAlias(givenAlias));

        // then
        assertThat(actual).isNotNull();
    }

    @Test
    void shouldPutPasswordEntry_InsertNewEntry() {
        // given
        final var existingKey = Bytes.bytesOf("Key");
        final var newKey = Bytes.bytesOf("tryThis");
        final var newPassword = Bytes.bytesOf("Password");
        final var matchingPasswordEntry = PasswordEntryFaker.faker()
            .fakePasswordEntry()
            .withKeyBytes(existingKey).fake();
        CryptoProviderFaker.faker()
            .forInstance(cryptoProvider)
            .withMockedEncryption().fake();
        PasswordEntryRepositoryFaker.faker()
            .forInstance(passwordEntryRepository)
            .withThesePasswordEntries(matchingPasswordEntry).fake();

        // when
        passwordService.putPasswordEntry(newKey, newPassword);

        // then
        then(cryptoProvider).should().encrypt(newKey);
        then(cryptoProvider).should().encrypt(newPassword);
        then(passwordEntryRepository).should().sync();
        then(passwordEntryRepository).should().add(eq(PasswordEntry.Companion.createPasswordEntry(DEFAULT, newKey, newPassword)));
        then(passbirdEventRegistry).should().processEvents();
    }

    @Test
    void shouldPutPasswordEntry_UpdateExistingEntry() {
        // given
        final var existingKey = Bytes.bytesOf("Key");
        final var newPassword = Bytes.bytesOf("Password");
        final var matchingPasswordEntry = PasswordEntryFaker.faker()
            .fakePasswordEntry()
            .withKeyBytes(existingKey).fake();
        CryptoProviderFaker.faker()
            .forInstance(cryptoProvider)
            .withMockedEncryption().fake();
        PasswordEntryRepositoryFaker.faker()
            .forInstance(passwordEntryRepository)
            .withThesePasswordEntries(matchingPasswordEntry).fake();

        // when
        passwordService.putPasswordEntry(existingKey, newPassword);

        // then
        then(cryptoProvider).should().encrypt(existingKey);
        then(cryptoProvider).should().encrypt(newPassword);
        then(passwordEntryRepository).should().sync();
        then(passbirdEventRegistry).should().processEvents();
        assertThatKeyExistsWithPassword(existingKey, newPassword);
    }

    @Test
    void shouldPutPasswordEntry_RejectInvalidKey() {
        // given
        final var invalidKey = Bytes.bytesOf("1Key");

        // when
        var actual = catchThrowable(() -> passwordService.putPasswordEntry(invalidKey, Bytes.bytesOf("password")));

        // then
        assertThat(actual).isNotNull().isInstanceOf(InvalidKeyException.class);
        then(cryptoProvider).shouldHaveNoInteractions();
        then(passwordEntryRepository).shouldHaveNoInteractions();
    }

    @Test
    void shouldPutPasswordEntries_InsertAndUpdate() {
        // given
        final var newKey = Bytes.bytesOf("trythis");
        final var newPassword = Bytes.bytesOf("dont use this as a password");
        final var existingKey = Bytes.bytesOf("Key");
        final var newPasswordForExistingKey = Bytes.bytesOf("Password");
        final var matchingPasswordEntry = PasswordEntryFaker.faker()
            .fakePasswordEntry()
            .withKeyBytes(existingKey).fake();
        CryptoProviderFaker.faker()
            .forInstance(cryptoProvider)
            .withMockedEncryption().fake();
        PasswordEntryRepositoryFaker.faker()
            .forInstance(passwordEntryRepository)
            .withThesePasswordEntries(matchingPasswordEntry).fake();

        // when
        passwordService.putPasswordEntries(Stream.of(
            new Tuple<>(newKey, newPassword),
            new Tuple<>(existingKey, newPasswordForExistingKey)
        ));

        // then
        then(cryptoProvider).should().encrypt(newKey);
        then(cryptoProvider).should().encrypt(existingKey);
        then(passwordEntryRepository).should().add(eq(PasswordEntry.Companion.createPasswordEntry(DEFAULT, newKey, newPassword)));
        then(passwordEntryRepository).should().sync();
        then(passbirdEventRegistry).should().processEvents();
        assertThatKeyExistsWithPassword(existingKey, newPasswordForExistingKey);
    }

    private void assertThatKeyExistsWithPassword(final Bytes keyBytes, final Bytes passwordBytes) {
        assertThat(passwordEntryRepository.find(keyBytes))
            .isNotEmpty().get()
            .extracting(PasswordEntry::viewPassword).isNotNull()
            .isEqualTo(passwordBytes);
    }

}

File: src/test/kotlin/de/pflugradts/passbird/domain/service/password/PutPasswordServiceTest.kt
package de.pflugradts.passbird.domain.service.password

import de.pflugradts.kotlinextensions.tryCatching
import de.pflugradts.passbird.application.security.fakeCryptoProvider
import de.pflugradts.passbird.domain.model.egg.Egg.Companion.createEgg
import de.pflugradts.passbird.domain.model.egg.InvalidEggIdException
import de.pflugradts.passbird.domain.model.egg.createEggForTesting
import de.pflugradts.passbird.domain.model.nest.NestSlot
import de.pflugradts.passbird.domain.model.shell.Shell.Companion.shellOf
import de.pflugradts.passbird.domain.model.shell.ShellPair
import de.pflugradts.passbird.domain.service.createNestServiceForTesting
import de.pflugradts.passbird.domain.service.eventhandling.EventRegistry
import de.pflugradts.passbird.domain.service.password.encryption.CryptoProvider
import de.pflugradts.passbird.domain.service.password.storage.EggRepository
import de.pflugradts.passbird.domain.service.password.storage.fakeEggRepository
import io.mockk.Called
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import strikt.api.expectThat
import strikt.assertions.isA
import strikt.assertions.isEqualTo
import strikt.assertions.isNotNull
import strikt.assertions.isTrue
import java.util.stream.Stream

class PutPasswordServiceTest {

    private val cryptoProvider = mockk<CryptoProvider>()
    private val eggRepository = mockk<EggRepository>(relaxed = true)
    private val eventRegistry = mockk<EventRegistry>(relaxed = true)
    private val nestService = createNestServiceForTesting()
    private val passwordService = PutPasswordService(cryptoProvider, eggRepository, eventRegistry, nestService)

    @Nested
    inner class ChallengeEggIdTest {
        @Test
        fun `should succeed when challenging alphabetic eggId`() {
            // given
            val givenEggId = shellOf("abcDEF")

            // when
            val actual = tryCatching { passwordService.challengeEggId(givenEggId) }

            // then
            expectThat(actual.success).isTrue()
        }

        @Test
        fun `should succeed when challenging eggId with digit at other than first position`() {
            // given
            val givenEggId = shellOf("abc123")

            // when
            val actual = tryCatching { passwordService.challengeEggId(givenEggId) }

            // then
            expectThat(actual.success).isTrue()
        }

        @Test
        fun `should fail when challenging eggId with digit at first position`() {
            // given
            val givenEggId = shellOf("123abc")

            // when
            val actual = tryCatching { passwordService.challengeEggId(givenEggId) }

            // then
            expectThat(actual.failure).isTrue()
            expectThat(actual.exceptionOrNull()).isNotNull().isA<InvalidEggIdException>()
        }

        @Test
        fun `should fail when challenging eggId with special characters`() {
            // given
            val givenEggId = shellOf("abc!")

            // when
            val actual = tryCatching { passwordService.challengeEggId(givenEggId) }

            // then
            expectThat(actual.failure).isTrue()
            expectThat(actual.exceptionOrNull()).isNotNull().isA<InvalidEggIdException>()
        }
    }

    @Test
    fun `should insert new egg`() {
        // given
        val existingEggId = shellOf("EggId")
        val newEggId = shellOf("tryThis")
        val newPassword = shellOf("Password")
        val matchingEgg = createEggForTesting(withEggIdShell = existingEggId)
        fakeCryptoProvider(instance = cryptoProvider)
        fakeEggRepository(instance = eggRepository, withEggs = listOf(matchingEgg))

        // when
        passwordService.putEgg(newEggId, newPassword)

        // then
        verify(exactly = 1) { cryptoProvider.encrypt(newEggId) }
        verify(exactly = 1) { cryptoProvider.encrypt(newPassword) }
        verify(exactly = 1) { eggRepository.sync() }
        verify(exactly = 1) { eggRepository.add(eq(createEgg(NestSlot.DEFAULT, newEggId, newPassword))) }
        verify(exactly = 1) { eventRegistry.processEvents() }
    }

    @Test
    fun `should update existing egg`() {
        // given
        val existingEggId = shellOf("EggId")
        val newPassword = shellOf("Password")
        val matchingEgg = createEggForTesting(withEggIdShell = existingEggId)
        fakeCryptoProvider(instance = cryptoProvider)
        fakeEggRepository(instance = eggRepository, withEggs = listOf(matchingEgg))

        // when
        passwordService.putEgg(existingEggId, newPassword)

        // then
        verify(exactly = 1) { cryptoProvider.encrypt(existingEggId) }
        verify(exactly = 1) { cryptoProvider.encrypt(newPassword) }
        verify(exactly = 1) { eggRepository.sync() }
        verify(exactly = 1) { eventRegistry.processEvents() }
        expectThat(eggRepository.find(eggIdShell = existingEggId).orElse(null).viewPassword()) isEqualTo newPassword
    }

    @Test
    fun `should reject invalid eggId`() {
        // given
        val invalidEggId = shellOf("1EggId")

        // when
        val actual = tryCatching { passwordService.putEgg(invalidEggId, shellOf("Password")) }

        // then
        expectThat(actual.failure).isTrue()
        expectThat(actual.exceptionOrNull()).isNotNull().isA<InvalidEggIdException>()
        verify { cryptoProvider wasNot Called }
        verify { eggRepository wasNot Called }
    }

    @Test
    fun `should upsert multiple eggs`() {
        // given
        val newEggId = shellOf("trythis")
        val newPassword = shellOf("dont use this as a password")
        val existingEggId = shellOf("EggId")
        val newPasswordForExistingEggId = shellOf("Password")
        val matchingEgg = createEggForTesting(withEggIdShell = existingEggId)
        fakeCryptoProvider(instance = cryptoProvider)
        fakeEggRepository(instance = eggRepository, withEggs = listOf(matchingEgg))

        // when
        passwordService.putEggs(
            Stream.of(ShellPair(newEggId, newPassword), ShellPair(existingEggId, newPasswordForExistingEggId)),
        )

        // then
        verify(exactly = 1) { cryptoProvider.encrypt(newEggId) }
        verify(exactly = 1) { cryptoProvider.encrypt(existingEggId) }
        verify(exactly = 1) { eggRepository.add(eq(createEgg(NestSlot.DEFAULT, newEggId, newPassword))) }
        verify(exactly = 1) { eggRepository.sync() }
        verify(exactly = 1) { eventRegistry.processEvents() }
        expectThat(
            eggRepository.find(eggIdShell = existingEggId).orElse(null).viewPassword(),
        ) isEqualTo newPasswordForExistingEggId
    }

    @Test
    fun `should accept empty stream`() {
        // given
        fakeCryptoProvider(instance = cryptoProvider)
        fakeEggRepository(instance = eggRepository)

        // when
        passwordService.putEggs(Stream.empty())

        // then
        verify(exactly = 0) { eggRepository.add(any()) }
        verify(exactly = 1) { eggRepository.sync() }
        verify(exactly = 1) { eventRegistry.processEvents() }
    }
}
